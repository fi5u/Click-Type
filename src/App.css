.App {
    background-color: #f1faee;
    display: flex;
    flex-direction: column;
    height: 100vh;
    justify-content: space-between;
    text-align: center;
}
