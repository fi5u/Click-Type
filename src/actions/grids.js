import * as types from './action-types'

export function select() {
    return {
        type: types.SELECT,
    }
}

export function updateSuggestedWords(words) {
    return {
        type: types.UPDATE_SUGGESTED_WORDS,
        words,
    }
}
