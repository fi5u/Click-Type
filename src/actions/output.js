import * as types from './action-types'

export function updateOutput(character) {
    return {
        type: types.UPDATE_OUTPUT,
        character,
    }
}
