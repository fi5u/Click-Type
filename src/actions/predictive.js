import * as types from './action-types'

export function addPredictiveWord(word, nextWord) {
    return {
        type: types.ADD_PREDICTIVE_WORD,
        word,
        nextWord,
    }
}
