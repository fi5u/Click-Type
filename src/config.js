const chars = {
    backspace: '⌫',
    backup: '⇡',
    capsLock: '⇪',
    space: '[]',
}

const punctuation = [
    chars.space,
    '.',
    ',',
    '’',
    '?',
    '!',
    '⇪',
]

const tightPunctuation = [
    // Should not have space before
    '.',
    ',',
    '’',
    '?',
    '!',
]

const suggestedWords = ['I', 'you', 'what', 'when']

const config = {
    capitalizedAfter: [
        '. ',
        '? ',
        '! ',
    ],
    chars: chars,
    gridParts: {
        additionals: {
            pre: [chars.backup, chars.backspace],
            post: [],
        },
        letters: [
            ['a', 'b', 'c', 'd', 'e'],
            ['f', 'g', 'h', 'i', 'j'],
            ['k', 'l', 'm', 'n', 'o'],
            ['p', 'q', 'r', 's', 't'],
            ['u', 'v', 'w', 'x', 'y', 'z']
        ],
        punctuation: [punctuation],
        suggestedWords: suggestedWords,
    },
    punctuation: punctuation,
    suggestedWordCount: 4,
    tickDuration: 400,
    tightPunctuation: tightPunctuation,
}

export { config }
