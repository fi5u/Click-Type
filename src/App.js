import './App.css'
import React, {
    Component,
} from 'react'
import {
    commonWords,
    words,
} from './data'
import {
    increaseSpeed,
    reduceSpeed,
} from './actions/settings'
import {
    select,
    setActiveColumn,
    toggleCapsLock,
    updateSuggestedWords,
} from './actions/grids'
import {
    setOutput,
    updateOutput,
} from './actions/output'
import {
    startTick,
    stopTick,
    tick,
} from './actions/timings'
import Grid from './components/Grid'
import Header from './components/Header'
import InfoBar from './components/InfoBar'
import LanguageProcessing from './services/language-processing'
import OutputDisplay from './components/OutputDisplay'
import PropTypes from 'prop-types'
import _ from 'lodash'
import { addPredictiveWord } from './actions/predictive'
import { config } from './config'
import { connect } from 'react-redux'

export class App extends Component { // export from here to allow tests w/out redux
    constructor() {
        super()

        this.state = {
            showClearConfirm: false,
        }

        document.addEventListener('keydown', this.detectClick.bind(this), true)
        this.addLodashMixins()
        this.langProcess = new LanguageProcessing()

        this.clickButton = this.clickButton.bind(this)
        this.clickMainButton = this.clickMainButton.bind(this)

        document.addEventListener('touchstart', event => {
            if(event.target.tagName !== 'BUTTON') {
                this.clickMainButton()
            }
        })
    }

    addLodashMixins() {
        _.mixin({
            'sortKeysBy': (obj, comparator) => {
                var keys = _.sortBy(_.keys(obj), (key) => {
                    return comparator ? comparator(obj[key], key) : key
                })

                return _.zipObject(keys, _.map(keys, (key) => {
                    return obj[key]
                }))
            }
        })
    }

    componentDidUpdate(prevProps) {
        if(prevProps.output === this.props.output) { return }
        this.props.dispatch(updateSuggestedWords(this.getSuggestedWords()))
    }

    clickButton({character, charType}, replace = false) {
        if(character === config.chars.clear) {
            if(this.state.showClearConfirm) {
                this.props.dispatch(setOutput(''))
                this.props.dispatch(toggleCapsLock(false))
                this.setState({
                    showClearConfirm: false,
                })
            }
            else {
                this.setState({
                    showClearConfirm: true,
                })
            }
            return
        }
        else if(this.state.showClearConfirm) {
            this.setState({
                showClearConfirm: false,
            })
        }

        if(character === config.chars.capsLock) {
            this.props.dispatch(toggleCapsLock(!this.props.settings.capsLock))
            return
        }

        if(character === config.chars.speedUp || character === config.chars.speedDown) {
            if((character === config.chars.speedDown && this.props.settings.canDecreaseSpeed) ||
                (character === config.chars.speedUp && this.props.settings.canIncreaseSpeed)) {
                this.stopAndPause(2)
                this.props.dispatch(character === config.chars.speedDown ? reduceSpeed() : increaseSpeed())
            }

            return
        }

        const isSuggestedWord = charType === 'suggested'
        this.props.dispatch(replace ? setOutput(character) : updateOutput(character, isSuggestedWord, this.props.settings))

        // Move focus back to first suggested word
        if(isSuggestedWord) {
            this.stopAndPause(2)
            this.props.dispatch(setActiveColumn(this.props.grid[this.props.activeRow].length))
        }

        // Save predictive words
        const outputWords = this.props.output.trim().split(' ')
        if((character.slice(-config.chars.space.length) === config.chars.space || character.slice(-1) === ' ' || isSuggestedWord) && outputWords.length > 1) {
            this.props.dispatch(addPredictiveWord(outputWords.concat(outputWords[outputWords.length - 1] !== character ? character : [])))
        }
    }

    clickMainButton() {
        if(this.props.tickStarted || this.props.activeAxis === 'col') {
            if(this.props.activeAxis === 'col') {
                this.clickButton(this.props.grid[this.props.activeRow].concat(this.props.suggestedWords)[this.props.activeElement])
            }
            this.props.dispatch(select())

            // Stop and pause
            this.stopAndPause()
        }
        else {
            this.startTick()
        }
    }

    detectClick(event) {
        // If space or typing directly in textarea, then ignore
        if(event.keyCode !== 32) { return }
        this.clickMainButton()
        event.preventDefault()
    }

    getPredictiveWords(words) {
        let foundWords = []
        let wordsLower = words.map(word => (word.toLowerCase()))
        if(_.isEmpty(words)) {
            foundWords = this.getSortedObj(this.props.predictiveWords)
        }
        let testWord = wordsLower[wordsLower.length - 1]
        if(testWord && testWord in this.props.predictiveWords) {
            let results = this.getSortedObj(this.props.predictiveWords[testWord].words)
            foundWords = results.length ? results : foundWords
        }

        if(wordsLower[wordsLower.length - 2]
            && wordsLower[wordsLower.length - 2] in this.props.predictiveWords
            && 'words' in this.props.predictiveWords[wordsLower[wordsLower.length - 2]]
            && wordsLower[wordsLower.length - 1] in this.props.predictiveWords[wordsLower[wordsLower.length - 2]].words
        ) {
            let results = this.getSortedObj(this.props.predictiveWords[wordsLower[wordsLower.length - 2]].words[wordsLower[wordsLower.length - 1]].words)
            foundWords = results.length ? results : foundWords
        }
        return foundWords
    }

    getSortedObj(obj) {
        return _.reverse(Object.keys(_.sortKeysBy(obj, value => { return value.freq })))
    }

    getSuggestedWords() {
        let output = this.props.output.toLowerCase()
        let suggestedWords = config.gridParts.suggestedWords
        const outputWords = output.trim().split(' ')
        const lastWord = outputWords.length > 0 ? outputWords[outputWords.length - 1] : ''

        if(outputWords.length > 0) {
            // Do not allow duplicates
            // Concat the default suggested words
            suggestedWords = _.uniqBy(this.getPredictiveWords(outputWords).concat(config.gridParts.suggestedWords), word => word.toLowerCase()).slice(0, this.props.grid.suggestedWordCount)

            // If endeded on a full word and sentence is a question, offer question mark
            if(output.slice(-1) === ' ') {
                const lastSentence = output.split(/\.|\?|!/g).pop()
                if(this.langProcess.shouldBeAQuestion(lastSentence)) {
                    suggestedWords = ['?'].concat(suggestedWords).slice(0, this.props.grid.suggestedWordCount)
                }
            }
        }
        if(output.trim().length > 0 && output.slice(-1) !== ' ') {
            // If last char is apostrope, suggest suitable replacements
            if(lastWord.slice(-1) === 'â€™' || lastWord.slice(-1) === '\'') {
                return this.langProcess.getApostrophizedWords(lastWord)
            }

            const wordPart = output.trim().split(' ').pop()
            // Get common words
            suggestedWords = this.getWordsFromArray(commonWords, wordPart, this.props.grid.suggestedWordCount)
            // If not enough common words, get from full dictionary
            if(suggestedWords.length < this.props.grid.suggestedWordCount) {
                const suggestedFullWords = this.getWordsFromArray(words, wordPart, this.props.grid.suggestedWordCount - suggestedWords.length, suggestedWords)
                suggestedWords = suggestedWords.concat(suggestedFullWords)
            }

            // If no suggested words, suggest a space plus default suggested words
            if(!suggestedWords.length) {
                suggestedWords = config.gridParts.suggestedWords.map(word => ` ${word}`)
            }
        }

        // An 'i' in suggested words should be capitalized
        for(let i = suggestedWords.length - 1; i >= 0; i--) {
            if(suggestedWords[i] === 'i') {
                suggestedWords[i] = 'I'
            }
        }

        // If has typed 'a', first suggested word should be a space
        if(lastWord === 'a' && output.slice(-1) !== ' ') {
            suggestedWords = [config.chars.space].concat(suggestedWords).slice(0, this.props.grid.suggestedWordCount)
        }

        return suggestedWords
    }

    getWordsFromArray(wordArray, match, count, ignoreValues = []) {
        const stringAtStart = [] // store words that contain string at start
        const stringInString = [] // store words that contain string not at start
        for(let i = 0, len = wordArray.length; i < len; i++) {
            if(ignoreValues.indexOf(wordArray[i]) > -1) {
                continue
            }
            if(wordArray[i].indexOf(match.toLowerCase()) === 0) {
                stringAtStart.push(wordArray[i])
            }
            else if(wordArray[i].indexOf(match.toLowerCase()) > 0) {
                stringInString.push(wordArray[i])
            }
            if(stringAtStart.length >= count) {
                break
            }
        }
        return stringAtStart.concat(stringInString).slice(0, count)
    }

    startTick() {
        this.props.dispatch(startTick())
        this.ticker = window.setInterval(() => {
            this.props.dispatch(tick())
        }, this.props.settings.speed)
    }

    stopAndPause(times = 1) {
        if(this.pauseInProgress)  { return }
        this.pauseInProgress = true
        this.stopTick()
        window.setTimeout(() => {
            this.pauseInProgress = false
            if(this.props.tickStarted) { return }
            this.startTick()
        }, this.props.settings.speed * times)
    }

    stopTick() {
        clearInterval(this.ticker)
        this.props.dispatch(stopTick())
    }

    render() {
        const {
            activeAxis,
            activeElement,
            activeRow,
            grid,
            isRunning,
            settings,
            suggestedWords,
            output,
        } = this.props

        return (
            <div
                className="App"
            >
                <Header
                    isRunning={isRunning}
                />

                <Grid
                    activeAxis={activeAxis}
                    activeElement={activeElement}
                    activeRow={activeRow}
                    characterGrid={grid}
                    clickButton={this.clickButton}
                    output={output}
                    settings={settings}
                    showClearConfirm={this.state.showClearConfirm}
                    suggestedWords={suggestedWords}
                />

                <OutputDisplay
                    value={output}
                />

                <InfoBar
                    settings={settings}
                />
            </div>
        )
    }
}

App.propTypes = {
    activeAxis: PropTypes.oneOf([
        'col',
        'row',
    ]).isRequired,
    activeElement: PropTypes.number.isRequired,
    activeRow: PropTypes.number.isRequired,
    dispatch: PropTypes.func.isRequired,
    grid: PropTypes.array.isRequired,
    isRunning: PropTypes.bool.isRequired,
    output: PropTypes.string.isRequired,
    predictiveWords: PropTypes.object.isRequired,
    settings: PropTypes.object.isRequired,
    suggestedWords: PropTypes.array.isRequired,
    tickStarted: PropTypes.bool.isRequired,
}

function mapStateToProps(state) {
    const {
        grids,
        output,
        predictive,
        settings,
        timings,
    } = state

    return {
        activeAxis: grids.activeAxis,
        activeElement: grids.activeElement,
        activeRow: grids.activeRow,
        grid: grids.activeGrid,
        isRunning: timings.isRunning,
        output: output.output,
        predictiveWords: predictive.words,
        settings,
        suggestedWords: grids.suggestedWords,
        tickStarted: timings.tickStarted,
    }
}

export default connect(mapStateToProps)(App)
