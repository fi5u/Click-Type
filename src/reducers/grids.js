import * as types from '../actions/action-types'
import { config } from '../config'

const addAdditionals = baseCharGroup => {
    const returnCharGroup = []
    for(const baseChars of baseCharGroup) {
        returnCharGroup.push(config.gridParts.additionals.pre.concat(baseChars, config.gridParts.additionals.post))
    }
    return returnCharGroup
}

// Index + 1 represents length for that number of suggested words
// i.e. 1 word can contain a max of 27 chars
// 2 words can contain a max of 13 chars
//const suggestedCounts = [27, 13, 7, 5]
const suggestedLetterCountMax = 20

export const initialState = {
    activeAxis: 'row',
    activeElement: 0,
    activeGrid: addAdditionals(config.gridParts.letters).concat(addAdditionals(config.gridParts.numbers)).concat(addAdditionals(config.gridParts.punctuation)),
    activeRow: 0,
    suggestedWords: config.gridParts.suggestedWords,
    suggestedWordCount: config.suggestedWordCountMax,
}

export default function grids(state = initialState, action) {
    switch(action.type) {

    case 'persist/REHYDRATE':
        return action.payload.settings && action.payload.settings.capsLock ? Object.assign({}, state, {
            activeGrid: addAdditionals(config.gridParts.letters).concat(addAdditionals(config.gridParts.numbers)).concat(addAdditionals(config.gridParts.secondaryPunc)),
        }) : state

    case types.SELECT:
        return Object.assign({}, state, {
            activeAxis: state.activeAxis === 'col' &&
                state.activeGrid[state.activeRow].concat(state.activeRow >= state.activeGrid.length - 2 ? [] : state.suggestedWords)[state.activeElement] === config.chars.backup ?
                'row' : 'col',
        })

    case types.SET_ACTIVE_COLUMN:
        return Object.assign({}, state, {
            activeElement: action.columnIndex,
        })

    case types.TICK:
        return Object.assign({}, state, {
            activeElement: state.activeAxis === 'row' ? 0 : state.activeElement >= state.activeGrid[state.activeRow].concat(state.activeRow >= state.activeGrid.length - 2 ? [] : state.suggestedWords).length - 1 ? 0 : state.activeElement + 1,
            activeRow: state.activeAxis === 'col' ? state.activeRow : state.activeRow >= state.activeGrid.length - 1 ? 0 : state.activeRow + 1,
        })

    case types.TOGGLE_CAPS_LOCK:
        return Object.assign({}, state, {
            activeGrid: addAdditionals(config.gridParts.letters).concat(addAdditionals(config.gridParts.numbers)).concat(addAdditionals(action.value ? config.gridParts.secondaryPunc : config.gridParts.punctuation)),
        })

    case types.UPDATE_SUGGESTED_WORDS: {
        let suggestedLetterCount = 0
        let wordCount = 0
        for(const word of action.words) {
            if(suggestedLetterCount + word.length > suggestedLetterCountMax || wordCount >= config.suggestedWordCountMax) {
                break
            }
            wordCount++
            suggestedLetterCount += word.length
        }
        return Object.assign({}, state, {
            suggestedWords: action.words.slice(0, wordCount),
            suggestedWordCount: wordCount,
        })
    }

    default:
        return state
    }
}
