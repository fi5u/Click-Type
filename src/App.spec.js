/* global expect, it, jest */
import * as gridsActions from './actions/grids'
import * as outputActions from './actions/output'
import * as predictiveActions from './actions/predictive'
import { mount, shallow } from 'enzyme'
import { App } from './App' // no Redux
import React from 'react'
import { config } from './config'
import renderer from 'react-test-renderer'

const props = {
    activeAxis: 'row',
    activeElement: 0,
    activeRow: 0,
    dispatch: jest.fn(),
    grid: [['a','b','c'],['d','e','f'],['g','h','i']],
    isRunning: false,
    output: '',
    predictiveWords: {},
    settings: {
        autoCapitalize: true,
        capsLock: false,
    },
    suggestedWords: ['abs', 'ace', 'adder'],
    tickStarted: false,
}

const predictiveWords = {
    apples: {
        freq: 2,
        words: {
            are: {
                freq: 1,
                words: {
                    green: {
                        freq: 1
                    },
                    healthy: {
                        freq: 3
                    },
                }
            },
            taste: {
                freq: 3,
                words: {
                    great: {
                        freq: 1
                    },
                    delicious: {
                        freq: 3
                    },
                    boring: {
                        freq: 2
                    },
                }
            }
        }
    },
    think: {
        freq: 4,
        words: {
            before: {
                freq: 1,
                words: {
                    we: {
                        freq: 1
                    },
                    I: {
                        freq: 2
                    },
                    you: {
                        freq: 4
                    },
                },
            },
            when: {
                freq: 3,
                words: {
                    you: {
                        freq: 4
                    },
                    tomorrow: {
                        freq: 1
                    },
                    is: {
                        freq: 6
                    },
                },
            },
            things: {
                freq: 2,
                words: {
                    through: {
                        freq: 2
                    },
                    are: {
                        freq: 1
                    },
                },
            },
        }
    }
}

it('renders App without crashing', () => {
    mount(
        <App {...props} />
    )
})

it('allows App props to be set', () => {
    const wrapper = mount(
        <App {...props} />
    )
    expect(wrapper.props().activeAxis).toEqual('row')
    wrapper.setProps({ activeAxis: 'col' })
    expect(wrapper.props().activeAxis).toEqual('col')
})

it('renders App correctly', () => {
    const tree = renderer.create(
        <App {...props} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

it('gets words from array correctly', () => {
    const app = new App()
    const inputTestArr = ['bat', 'ball', 'car']
    expect(app.getWordsFromArray(inputTestArr, 'ba', 10)).toEqual(['bat', 'ball'])
    expect(app.getWordsFromArray(inputTestArr, 'Ba', 10)).toEqual(['bat', 'ball'])
    expect(app.getWordsFromArray(inputTestArr, 'ba', 1)).toEqual(['bat'])
    expect(app.getWordsFromArray(inputTestArr, 'a', 3)).toEqual(inputTestArr)
    expect(app.getWordsFromArray(inputTestArr, 'a', 3, ['bat', 'ball'])).toEqual(['car'])
})

it('sorts an object', () => {
    const wrapper = mount(
        <App {...props} />
    )

    let sortedWords = wrapper.instance().getSortedObj(predictiveWords.apples.words.are.words)
    expect(sortedWords[0]).toBe('healthy')
    expect(sortedWords[1]).toBe('green')

    sortedWords = wrapper.instance().getSortedObj(predictiveWords.think.words.before.words)
    expect(sortedWords[0]).toBe('you')
    expect(sortedWords[1]).toBe('I')
    expect(sortedWords[2]).toBe('we')
})

it('gets predictive words', () => {
    const wrapper = mount(
        <App {...props} predictiveWords={predictiveWords} />
    )
    let predicitedWords = wrapper.instance().getPredictiveWords([])
    expect(predicitedWords[0]).toBe('think')
    expect(predicitedWords[1]).toBe('apples')

    predicitedWords = wrapper.instance().getPredictiveWords(['think'])
    expect(predicitedWords[0]).toBe('when')
    expect(predicitedWords[1]).toBe('things')
    expect(predicitedWords[2]).toBe('before')

    predicitedWords = wrapper.instance().getPredictiveWords(['Think'])
    expect(predicitedWords[0]).toBe('when')
    expect(predicitedWords[1]).toBe('things')
    expect(predicitedWords[2]).toBe('before')

    predicitedWords = wrapper.instance().getPredictiveWords(['think', 'when'])
    expect(predicitedWords[0]).toBe('is')
    expect(predicitedWords[1]).toBe('you')
    expect(predicitedWords[2]).toBe('tomorrow')

    predicitedWords = wrapper.instance().getPredictiveWords(['Think', 'when'])
    expect(predicitedWords[0]).toBe('is')
    expect(predicitedWords[1]).toBe('you')
    expect(predicitedWords[2]).toBe('tomorrow')

    predicitedWords = wrapper.instance().getPredictiveWords(['what', 'do' , 'you', 'think', 'when'])
    expect(predicitedWords[0]).toBe('is')
    expect(predicitedWords[1]).toBe('you')
    expect(predicitedWords[2]).toBe('tomorrow')

    predicitedWords = wrapper.instance().getPredictiveWords(['I', 'think'])
    expect(predicitedWords[0]).toBe('when')
    expect(predicitedWords[1]).toBe('things')
    expect(predicitedWords[2]).toBe('before')

    predicitedWords = wrapper.instance().getPredictiveWords(['do', 'when'])
    expect(predicitedWords.length).toBe(0)

    predicitedWords = wrapper.instance().getPredictiveWords(['truck'])
    expect(predicitedWords.length).toBe(0)
})

it('gets suggested words', () => {
    const wrapper = mount(
        <App {...props} output="ca" predictiveWords={predictiveWords} />
    )
    let suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0].indexOf('ca') > -1).toBe(true)
    wrapper.setProps({ output: 'br' })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0].indexOf('br') > -1).toBe(true)
    wrapper.setProps({
        output: 'I think ',
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe('when')
    expect(suggestedWords[1]).toBe('things')
    expect(suggestedWords[2]).toBe('before')

    wrapper.setProps({
        output: 'Apples are ',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe(config.gridParts.suggestedWords[0])
    expect(suggestedWords[1]).toBe(config.gridParts.suggestedWords[1])
    expect(suggestedWords[2]).toBe(config.gridParts.suggestedWords[2])

    wrapper.setProps({
        output: 'What do ',
        predictiveWords: {
            what: {
                freq: 1,
                words: {
                    do: {
                        freq: 1,
                        words: {
                            i: {
                                freq: 1,
                            }
                        }
                    }
                }
            }
        },
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe('I')
    expect(suggestedWords[1]).not.toBe('I')
    expect(suggestedWords[2]).not.toBe('I')

    wrapper.setProps({
        output: 'dad’',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords.indexOf('dad’s')).toBeGreaterThan(-1)

    wrapper.setProps({
        output: 'dad\'',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords.indexOf('dad\'s')).toBeGreaterThan(-1)

    wrapper.setProps({
        output: 'it’',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords.indexOf('it’s')).toBeGreaterThan(-1)
    expect(suggestedWords.indexOf('it’ll')).toBeGreaterThan(-1)

    wrapper.setProps({
        output: 'Apples ',
        predictiveWords: predictiveWords,
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe('taste')

    wrapper.setProps({
        output: 'What is your name ',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe('?')

    wrapper.setProps({
        output: 'Hello you.',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe(' I')
    expect(suggestedWords[1]).toBe(' you')

    wrapper.setProps({
        output: 'Hello you!',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe(' I')
    expect(suggestedWords[1]).toBe(' you')

    wrapper.setProps({
        output: 'a',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).toBe(config.chars.space)
    expect(suggestedWords[1]).toBe('and')

    wrapper.setProps({
        output: 'a ',
        predictiveWords: {},
    })
    suggestedWords = wrapper.instance().getSuggestedWords()
    expect(suggestedWords[0]).not.toBe(config.chars.space)
})

it('handles the click event', () => {
    const wrapper = shallow(
        <App {...props} output="ca" predictiveWords={predictiveWords} />
    )
    let output = 'canary'
    outputActions.setOutput = jest.fn()
    outputActions.updateOutput = jest.fn()
    predictiveActions.addPredictiveWord = jest.fn()

    wrapper.instance().clickButton(output)
    expect(outputActions.updateOutput).toHaveBeenCalledWith('canary', wrapper.instance().props.suggestedWords.indexOf(output) > -1, wrapper.instance().props.settings)

    wrapper.setProps({
        output: 'I love my ',
        suggestedWords: ['canary'],
    })
    wrapper.instance().clickButton(output)
    expect(predictiveActions.addPredictiveWord).toHaveBeenCalledWith(['I', 'love', 'my', 'canary'])

    output = 'I love my canary'
    wrapper.setProps({
        output: 'I love my ',
        suggestedWords: [],
    })
    wrapper.instance().clickButton(output, true)
    expect(outputActions.setOutput).toHaveBeenCalledWith(output)
})

it('handles the click event by moving to the first suggested word after selecting a suggested word', () => {
    const wrapper = mount(
        <App {...props}
            activeElement={5}
            output="ad"
            predictiveWords={predictiveWords}
        />
    )
    let output = 'adder'
    gridsActions.setActiveColumn = jest.fn()
    wrapper.instance().clickButton(output)
    expect(gridsActions.setActiveColumn).toHaveBeenCalledWith(wrapper.props().grid[wrapper.props().activeRow].length)
})

it('shows clear confirmation when clear button pressed', () => {
    const wrapper = mount(
        <App {...props} />
    )
    expect(wrapper.state('showClearConfirm')).toEqual(false)
    wrapper.instance().clickButton(config.chars.clear)
    expect(wrapper.state('showClearConfirm')).toEqual(true)
})

it('clears output text when clear confirm pressed', () => {
    const wrapper = mount(
        <App {...props}
            output="Some text here"
        />
    )
    outputActions.setOutput = jest.fn()
    wrapper.setState({ showClearConfirm: true })
    wrapper.instance().clickButton(config.chars.clear)
    expect(wrapper.state('showClearConfirm')).toEqual(false)
    expect(outputActions.setOutput).toHaveBeenCalledWith('')
})
