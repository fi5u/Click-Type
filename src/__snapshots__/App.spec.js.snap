// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App correctly 1`] = `
<div
  className="App"
>
  <div
    className="ui borderless Header menu"
  >
    <a
      className="item"
      href="/"
      onClick={[Function]}
    >
      ClickType
    </a>
    <div
      className="right menu"
    >
      <div
        className="disabled item"
        onClick={[Function]}
      >
        Press space to start
      </div>
      <a
        className="item"
        onClick={[Function]}
      >
        Link Puck
        <i
          aria-hidden="true"
          className="help circle outline icon Header--question"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
            }
          }
        />
      </a>
      <a
        className="item"
        href="https://github.com/fi5u/clicktype/wiki/ClickType-help"
        onClick={[Function]}
      >
        Help
      </a>
    </div>
  </div>
  <div
    className="Grid"
  >
    <div
      className="ui basic buttons GridRow GridRow--is-active GridRow--is-first"
    >
      <button
        className="ui medium button GridItem GridItem--letter     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        a
      </button>
      <button
        className="ui medium button GridItem GridItem--letter     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        b
      </button>
      <button
        className="ui medium button GridItem GridItem--letter     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        c
      </button>
      <button
        className="ui medium button GridItem GridItem--suggested     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        abs
      </button>
      <button
        className="ui medium button GridItem GridItem--suggested     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        ace
      </button>
      <button
        className="ui medium button GridItem GridItem--suggested     GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        adder
      </button>
    </div>
    <div
      className="ui basic buttons GridRow  GridRow--is-middle"
    >
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        d
      </button>
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        e
      </button>
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        f
      </button>
    </div>
    <div
      className="ui basic buttons GridRow  GridRow--is-last"
    >
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        g
      </button>
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        h
      </button>
      <button
        className="ui medium button GridItem GridItem--letter  GridItem--is-inactive-row   GridItem--is-capitalized"
        disabled={undefined}
        onClick={[Function]}
        tabIndex={undefined}
      >
        i
      </button>
    </div>
  </div>
  <div
    className="OutputDisplay"
  >
    <span
      className="OutputDisplay__content"
    >
      
    </span>
  </div>
  <div
    className="InfoBar"
    style={undefined}
  >
    <div
      className="InfoBar__inner"
    >
      <div
        className="InfoBar__items"
      >
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              ⇡
            </div>
            <div
              className="InfoBar__label"
            >
              switch to rows
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              ⌫
            </div>
            <div
              className="InfoBar__label"
            >
              backspace
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              []
            </div>
            <div
              className="InfoBar__label"
            >
              space
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              «
            </div>
            <div
              className="InfoBar__label"
            >
              slow down
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              NaN
            </div>
            <div
              className="InfoBar__label"
            >
              current speed
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              »
            </div>
            <div
              className="InfoBar__label"
            >
              speed up
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              ⇪
            </div>
            <div
              className="InfoBar__label"
            >
              toggle caps lock
            </div>
          </div>
        </div>
        <div
          className="InfoBar__item"
        >
          <div
            className="InfoBar__item-inner"
          >
            <div
              className="InfoBar__symbol"
            >
              ×
            </div>
            <div
              className="InfoBar__label"
            >
              clear all
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
